dids/src/main/kotlin/web5/sdk/dids/DidDht.kt

```diff
@@ -0,0 +1,424 @@
+        }
+
+        message.addRecord(
+          TXTRecord(
+            Name("_$vmId._did."),
+            DClass.IN,
+            TTL,
+            "id=${verificationMethod.id.rawFragment};t=$keyType;k=$base64UrlEncodedKey"
+          ), Section.ANSWER
+        )
+
+        verificationMethodIds += vmId
+      }
+
+      // Add Resource Records for each Service
+      didDocument.services?.forEachIndexed { i, service ->
+        val sId = "s$i"
+        message.addRecord(
+          TXTRecord(
+            Name("_$sId._did."),
+            DClass.IN,
+            TTL,
+            "id=${service.id.rawFragment};t=${service.type};uri=${service.serviceEndpoint}"
+          ), Section.ANSWER
+        )
+        serviceIds += sId
+      }
+
+      // Construct top-level Resource Record
+      val rootRecordText = mutableListOf<String>().apply {
+        if (verificationMethodIds.isNotEmpty()) add("vm=${verificationMethodIds.joinToString(",")}")
+        if (serviceIds.isNotEmpty()) add("svc=${serviceIds.joinToString(",")}")
+        didDocument.authenticationVerificationMethodsDereferenced?.
+        map { verificationMethodsById[it.id.toString()] }?.joinToString(",")?.let { add("auth=$it") }
+        didDocument.assertionMethodVerificationMethodsDereferenced?.
+        map { verificationMethodsById[it.id.toString()] }?.joinToString(",")?.let { add("asm=$it") }
+        didDocument.keyAgreementVerificationMethodsDereferenced?.
+        map { verificationMethodsById[it.id.toString()] }?.joinToString(",")?.let { add("agm=$it") }
+        didDocument.capabilityInvocationVerificationMethodsDereferenced?.
+        map { verificationMethodsById[it.id.toString()] }?.joinToString(",")?.let { add("inv=$it") }
+        didDocument.capabilityDelegationVerificationMethodsDereferenced?.
+        map { verificationMethodsById[it.id.toString()] }?.joinToString(",")?.let { add("del=$it") }
+      }
+
+      message.addRecord(
+        TXTRecord(
+          Name("_did."), DClass.IN, TTL, rootRecordText.joinToString(";")
+        ), Section.ANSWER
+      )
+
+      // if there are types, add a Resource Record for them
+      if (types != null) {
+        message.addRecord(
+          TXTRecord(
+            Name("_typ._did."), DClass.IN, TTL, "id=${types.joinToString(",")}"
+          ), Section.ANSWER
+        )
+      }
+
+      return message
+    }
+
+    /**
+     * Converts a DNS packet to a [DIDDocument] according to the did:dht spec
+     * https://tbd54566975.github.io/did-dht-method/#dids-as-a-dns-packet
+     *
+     * @param did The DID that the packet is for.
+     * @param msg The [Message] instance containing the DNS packet.
+     * @return A [Pair] containing the [DIDDocument] and a list of types.
+     * @throws IllegalArgumentException if the provided DID does not conform to the "did:dht" method.
+     */
+    public fun fromDNSPacket(did: String, msg: Message): Pair<DIDDocument, List<Int>> {
```

Comment:
🙈 apologies in advance for saying: `fromDnsPacket` to remain consistent with convention in this repo